feat: Create a screen listing all attributes for a product

Sudhanva-Nadiger opened the pull request:
### What
- I have just added a simple button in edit product page just to test the new page. It will be removed and added in the paticular screen.
- taking `edit_product_page.dart` as a base `product_attribute_page.dart` is designed as suggested by teolemon.
- Added all the mentioned section field in the issue.

### Screenshot

<img src="https://github.com/openfoodfacts/smooth-app/assets/93595710/8509fc24-d0dd-4081-84c8-65d2290c6fe5" width="300" />

### Part of 
- #4673

Sudhanva-Nadiger committed:
feat: Create a screen listing all attributes for a product

- resolves: #4673
- I have just added a simple button in edit product page just to test the
new page. It will be removed and added in the paticular screen.
- taking `edit_product_page.dart` as a base `product_attribute_page.dart`
is designed as suggested by teolemon.
- Added all the mentioned section field in the issue.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/edit_product_page.dart:
```diff
@@ -448,3 +374,38 @@ class _ProductBarcodeState extends State<_ProductBarcode> {
   }
 }
+
+/// SVG that looks like a ListTile icon.
```
Please create a specific new file.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/edit_product_page.dart:
```diff
@@ -254,6 +200,18 @@ class _EditProductPageState extends State<EditProductPage> with UpToDateMixin {
                 },
               ),
+              ElevatedButton(
+                onPressed: () {
+                  Navigator.push(
```
`async`

monsieurtanuki commented on packages/smooth_app/lib/pages/product/edit_product_page.dart:
```diff
@@ -254,6 +200,18 @@ class _EditProductPageState extends State<EditProductPage> with UpToDateMixin {
                   );
                 },
               ),
+              ElevatedButton(
```
Please put it as optional, using a new dev mode flag (false by default).

monsieurtanuki commented on packages/smooth_app/lib/pages/product/common/product_app_bar.dart:
```diff
@@ -0,0 +1,92 @@
+      {required this.barcodeVisibleInAppbar,
+      required this.productName,
+      required this.productBrand,
+      required this.barcode});
```
Please put a comma after the last parameter.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/common/product_app_bar.dart:
```diff
@@ -0,0 +1,92 @@
+  final bool barcodeVisibleInAppbar;
+
+  @override
+  Size get preferredSize => const Size.fromHeight(kToolbarHeight);
```
Not sure about this one, as I couldn't find it in the original code.

g123k replied:
+1 you don't need a custom AppBar here, it will be in a tab in the new design.

Sudhanva-Nadiger replied:
Since its implementing `PreferredSizeWidget` !
Initially it was directly in the `edit_product_page.dart`. Now its on seperate file ! and appBar specifically takes preferredsizewidget! so had to add that .

monsieurtanuki commented on packages/smooth_app/lib/pages/product/common/product_app_bar.dart:
```diff
@@ -0,0 +1,92 @@
+      required this.productBrand,
+      required this.barcode});
+
+  final String productName;
+  final String productBrand;
+  final String barcode;
```
Please just put the product itself as parameter, and not 3 parameters derived from the product.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_attributes_page.dart:
```diff
@@ -0,0 +1,344 @@
+import 'package:smooth_app/pages/product/simple_input_page_helpers.dart';
+import 'package:smooth_app/widgets/smooth_scaffold.dart';
+
+const String splitChar = ':';
```
Please make it private.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_attributes_page.dart:
```diff
@@ -0,0 +1,344 @@
+class _ProductAttributesPageState extends State<ProductAttributesPage>
+    with UpToDateMixin {
+  final ScrollController _controller = ScrollController();
+  // OrderedNutrientsCache? cache;
```
Please get rid of useless comments.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_attributes_page.dart:
```diff
@@ -0,0 +1,344 @@
+                helper: SimpleInputPageLabelHelper(),
+                product: upToDateProduct,
+              ),
+              NutritionFacts(
```
Too much possible confusion with such a vague name.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_attributes_page.dart:
```diff
@@ -0,0 +1,344 @@
+                helper: SimpleInputPageCategoryHelper(),
+                product: upToDateProduct,
+              ),
+              AttributeItems(
```
Too much possible confusion with such a vague name.

Sudhanva-Nadiger replied:
can you suggest any other name pls! :p
 Coming up with good variableName is a nightmare xD

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_attributes_page.dart:
```diff
@@ -0,0 +1,344 @@
+              NutritionFacts(
+                product: upToDateProduct,
+              ),
+              Ingredients(
```
Too much possible confusion with such a vague name.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_attributes_page.dart:
```diff
@@ -0,0 +1,344 @@
+
+  @override
+  Widget build(BuildContext context) {
+    widget.product.knowledgePanels?.panelIdToPanelMap['nutrition_facts_table']
```
I hope there's an easier way to get nutrition facts.

g123k replied:
+ it would be better if this would be computed only once and not for every build

Sudhanva-Nadiger replied:
yeah! `product.nutriments`
But problem with that is the way its configured! so thinking that the names can have any symbol i just used this method which gives a map of key value pair !
So what should be done ?

monsieurtanuki replied:
@Sudhanva-Nadiger We should ignore the knowledge panels here, and loop on the nutrients (ordered according to the country), displaying whatever is populated.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_attributes_page.dart:
```diff
@@ -0,0 +1,344 @@
+}
+
+class _IngredientsState extends State<Ingredients> {
+  List<String> allIngredients = <String>[];
```
Should be private.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_attributes_page.dart:
```diff
@@ -0,0 +1,344 @@
+  Widget build(BuildContext context) {
+    widget.product.ingredients?.forEach((Ingredient element) {
+      if (element.text != null) {
+        allIngredients.add(element.text!);
```
You have to clear the List before, haven't you?

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_attributes_page.dart:
```diff
@@ -0,0 +1,344 @@
+  }
+}
+
+class _ListTile extends StatefulWidget {
```
Please put it in a specific file, and name it with a less vague name, like `AttributeFirstRowWidget`.

monsieurtanuki reviewed (commented):
Hi @Sudhanva-Nadiger, and thank you for your PR!
I have many comments: please have a look at them. And fix the 2 lint check errors, too.
I'll have more comments later, but I need you to refactor the code first.

g123k commented on packages/smooth_app/lib/pages/product/product_attributes_page.dart:
```diff
@@ -0,0 +1,344 @@
+        productBrand: productBrand,
+        productName: productName,
+      ),
+      body: RefreshIndicator(
```
There's no need to manually refresh this tab

g123k commented on packages/smooth_app/lib/pages/product/product_attributes_page.dart:
```diff
@@ -0,0 +1,344 @@
+    return _ListTile(
+      allTerms: allTerms,
+      leading: const SvgIcon('assets/cacheTintable/scale-balance.svg',
+          dontAddColor: true),
```
Please add a trailing comma

g123k commented on packages/smooth_app/lib/pages/product/product_attributes_page.dart:
```diff
@@ -0,0 +1,344 @@
+        for (int i = 0; i < row.values.length - 1; i++) {
+          term += row.values[i].text;
+          if (i == 0) {
+            term += splitChar;
```
Please use a `StringBuffer` when concatenating strings.

Sudhanva-Nadiger replied:
> it would be better if this would be computed only once and not for every build

I cant access `widget.product` above the `build` func

g123k replied:
Mmm, false, you can access to any part of the widget once it's `mounted`.

Sudhanva-Nadiger replied:
hi @g123k ! i have moved this logic to initstate so it should be fine now ! pls have a look at it!

g123k commented on packages/smooth_app/lib/pages/product/product_attributes_page.dart:
```diff
@@ -0,0 +1,344 @@
+  }
+
+  Widget _termsList(List<String> terms,
+      {bool hasTrailing = false, bool borderFlag = false}) {
```
Please add a trailing comma

monsieurtanuki commented:
https://github.com/openfoodfacts/smooth-app/issues/4673#issuecomment-1764229538

@g123k Next time please don't involve me in an issue if you're already reviewing the PR yourself; I have already enough work to do.
@Sudhanva-Nadiger I switch to other tasks so feel free to ignore my comments.

Sudhanva-Nadiger committed:
refactor: chnages done according to the feedback

- created seperte file for `svg_icon.dar`
- created newFile for `attribute_first_row_widget.dart`
- removed refreshIndicator
- useed stringbuffer for string concatenation
- fix: linting errors

g123k commented:
> [#4673 (comment)](https://github.com/openfoodfacts/smooth-app/issues/4673#issuecomment-1764229538)
> 
> @g123k Next time please don't involve me in an issue if you're already reviewing the PR yourself; I have already enough work to do. @Sudhanva-Nadiger I switch to other tasks so feel free to ignore my comments.

Oops, I'm really sorry, for me, my comments were complementary to yours.
@Sudhanva-Nadiger Please follow the comments from @monsieurtanuki. Once it's OK, it will give you my feedback about the UI.

Sudhanva-Nadiger commented:
Yeah Sure ! I have pushed the new commit ! @g123k

Sudhanva-Nadiger committed:
fix: move nutrients extraction to init state

- to avoid recalculation on every build

M123-dev commented:
@monsieurtanuki what do you think code wise

monsieurtanuki commented on packages/smooth_app/lib/pages/product/product_attributes_page.dart:
```diff
@@ -0,0 +1,243 @@
+            ProductAttributeNutritionFacts(
+              product: upToDateProduct,
+            ),
+            ProductAtrributeIngredients(
```
typo

monsieurtanuki reviewed (commented):
Hi @Sudhanva-Nadiger!
I'm sorry I couldn't follow your code changes. I'm more available now.
Could you please make sure that the tests are OK: https://github.com/openfoodfacts/smooth-app/actions/runs/6615121230/job/17966602417?pr=4714. That's the first step.

Sudhanva-Nadiger committed:
remove usage of knowledgepanel, use product.nutriments

Sudhanva-Nadiger committed:
move ingredient extraction to initstate

monsieurtanuki commented on packages/smooth_app/lib/pages/product/attribute_first_row_widget.dart:
```diff
@@ -0,0 +1,116 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_gen/gen_l10n/app_localizations.dart';
+
+const String _SplitChar = ':';
```
We don't need it.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/attribute_first_row_widget.dart:
```diff
@@ -0,0 +1,116 @@
+
+  final Widget? leading;
+  final String title;
+  final List<String> allTerms;
```
List<StringPair> instead.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/attribute_first_row_widget.dart:
```diff
@@ -0,0 +1,116 @@
+    required this.onTap,
+  });
+
+  final Widget? leading;
```
```suggestion
  final Widget leading;
```

monsieurtanuki commented on packages/smooth_app/lib/pages/product/attribute_first_row_widget.dart:
```diff
@@ -0,0 +1,116 @@
+  final Widget? leading;
+  final String title;
+  final List<String> allTerms;
+  final bool hasTrailing;
```
We don't need it.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/attribute_first_row_widget.dart:
```diff
@@ -0,0 +1,116 @@
+  final String title;
+  final List<String> allTerms;
+  final bool hasTrailing;
+  final Function()? onTap;
```
```suggestion
  final Function() onTap;
```

monsieurtanuki commented on packages/smooth_app/lib/pages/product/attribute_first_row_widget.dart:
```diff
@@ -0,0 +1,116 @@
+}
+
+class _AttributeFirstRowWidgetState extends State<AttributeFirstRowWidget> {
+  bool showAllTerms = false;
```
```suggestion
  bool _showAllTerms = false;
```

monsieurtanuki commented on packages/smooth_app/lib/pages/product/attribute_first_row_widget.dart:
```diff
@@ -0,0 +1,116 @@
+  Widget build(BuildContext context) {
+    final AppLocalizations appLocalizations = AppLocalizations.of(context);
+    final ThemeData theme = Theme.of(context);
+    final bool hasMoreThanFourTerms = widget.allTerms.length > 4;
+    final List<String> firstFourItems = widget.allTerms.take(4).toList();
```
```suggestion
    const int numberThreshold = 4;
    final bool hasManyTerms = widget.allTerms.length > numberThreshold;
    final List<String> firstTerms = widget.allTerms.take(numberThreshold).toList();
```

monsieurtanuki commented on packages/smooth_app/lib/pages/product/attribute_first_row_widget.dart:
```diff
@@ -0,0 +1,116 @@
+          hasTrailing: widget.hasTrailing,
+          borderFlag: !hasMoreThanFourTerms,
+        ),
+        Column(
```
Don't put `Column` inside another `Column`.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/attribute_first_row_widget.dart:
```diff
@@ -0,0 +1,116 @@
+                  bottom: BorderSide(),
+                ),
+          trailing:
+              hasTrailing ? Text(terms[index].split(_SplitChar)[1]) : null,
```
Instead something like `trailing: terms[index].second == null ? null : Text(terms[index].second!)`

monsieurtanuki commented on packages/smooth_app/lib/pages/product/attribute_first_row_widget.dart:
```diff
@@ -0,0 +1,116 @@
+    );
+  }
+
+  Widget _termsList(
```
I would return a `List<Widget>` instead of a `ListView` that itself is going to be in a `Column`.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/attribute_first_row_widget.dart:
```diff
@@ -0,0 +1,116 @@
+            if (hasMoreThanFourTerms) ...<Widget>[
+              if (showAllTerms) ...<Widget>[
+                _termsList(
+                  widget.allTerms.skip(firstFourItems.length).toList(),
```
I don't think we should skip the first items. We should just display them all.

monsieurtanuki reviewed (commented):
Hi @Sudhanva-Nadiger!

Thank you for your work!

Unfortunately you've put too much effort on something we actually don't need regarding UI. Please rollback
* edit_product_page.dart
* product_app_bar.dart
* product_attributes_page.dart
while keeping `svg_icon.dart` in a specific file as "not private anymore" (which makes sense).

Please create a new class `StringPair` with `final String first` and `final String? second` fields.
To be used instead of pasting 2 strings with a separator, as luckily we're coding in OOP and not in shell.

Please have a look at my comments for attribute_first_row_widget.dart.

Please create the file `attribute_first_row_helper.dart` with
* abstract class `AttributeFirstRowHelper` in it, with method `List<StringPair> getTerms();`
* class `AttributeFirstRowSimpleHelper`, that takes an `AbstractSimpleInputPageHelper` in the constructor
* class `AttributeFirstRowNutritionHelper`, that takes the `Product` in the constructor
* class `AttributeFirstRowIngredientsHelper`, that takes the `Product` in the constructor

With that you may say: "hey, we're not displaying anything!". That's correct. We'll deal with that once you're done with my suggestions.

Sudhanva-Nadiger commented:
@monsieurtanuki i will abstract all the things we need in firstRowwidget! like `allTerms`, `leadng`, `title` and `onTap`! whats ur suggestion?

Sudhanva-Nadiger committed:
helper abstract class for attribute_first_row_widget

monsieurtanuki commented on packages/smooth_app/lib/pages/product/attribute_first_row_widget.dart:
```diff
@@ -0,0 +1,115 @@
+
+class _AttributeFirstRowWidgetState extends State<AttributeFirstRowWidget> {
+  bool showAllTerms = false;
+  late List<StringPair> allTerms = <StringPair>[];
```
```suggestion
  late final List<StringPair> allTerms;
```

monsieurtanuki commented on packages/smooth_app/lib/pages/product/atrribute_first_row_helper.dart:
```diff
@@ -0,0 +1,138 @@
+import 'package:smooth_app/pages/product/simple_input_page_helpers.dart';
+
+class StringPair {
+  StringPair({
```
```suggestion
  const StringPair({
```

monsieurtanuki commented on packages/smooth_app/lib/pages/product/atrribute_first_row_helper.dart:
Typo in the file name.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/attribute_first_row_widget.dart:
```diff
@@ -0,0 +1,115 @@
+      shrinkWrap: true,
+      itemBuilder: (_, int index) {
+        return ListTile(
+          key: UniqueKey(),
```
Do we need a key?

monsieurtanuki reviewed (commented):
Thank you @Sudhanva-Nadiger, it's much better now!

My main new comment would be that in `AttributeFirstRowHelper` you should add something like that:
```dart
Future<void> onTap({
  required final BuildContext context,
  required final Product product,
})
```

In addition to that:
* please remove `product_app_bar.dart`
* please remove `product_attributes_page.dart`
* please rollback `edit_product_page.dart`

This PR will focus on building the tools (basically `AttributeFirstRowHelper` and `AttributeFirstRowWidget`), without display.
If you don't mind once this PR is approved and merged we'll focus on where and when we should display the data.

Sudhanva-Nadiger commented:
Ok cool I will do that 👍🏻!

Sudhanva-Nadiger committed:
remove ui file and abstract ontap function

Sudhanva-Nadiger committed:
tiny fix

monsieurtanuki commented on packages/smooth_app/lib/pages/product/attribute_first_row_helper.dart:
```diff
@@ -0,0 +1,202 @@
+  Function() onTap({
+    required BuildContext context,
+    required Product product,
+    bool mounted = false,
```
You're missing the point: if needed, inside the method, you may call `context.mounted`. But `mounted` is not a parameter from outside.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/attribute_first_row_helper.dart:
```diff
@@ -0,0 +1,202 @@
+
+  String getTitle(BuildContext context);
+
+  Function() onTap({
```
I'd rather say `Future<void> onTap` if you don't mind.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/attribute_first_row_helper.dart:
```diff
@@ -0,0 +1,202 @@
+
+  Function() onTap({
+    required BuildContext context,
+    required Product product,
```
Actually we don't need a `Product` here: we already have one in all implementations (helper + nutrition + ingredients).
`Future<void> onTap(final BuildContext context)` will really look better.

monsieurtanuki commented on packages/smooth_app/lib/pages/product/attribute_first_row_widget.dart:
```diff
@@ -0,0 +1,119 @@
+class AttributeFirstRowWidget extends StatefulWidget {
+  const AttributeFirstRowWidget({
+    required this.helper,
+    required this.product,
```
We won't need `product` here.

monsieurtanuki reviewed (commented):
@Sudhanva-Nadiger We're getting closer!
Please have a look at my comments, and we'll probably be good for an approved PR ;)

Sudhanva-Nadiger committed:
change ontap function to future<void>

Sudhanva-Nadiger commented:
@monsieurtanuki done !

monsieurtanuki commented on packages/smooth_app/lib/pages/product/attribute_first_row_helper.dart:
```diff
@@ -118,34 +115,30 @@ class AttributeFirstRowNutritionHelper extends AttributeFirstRowHelper {
+      return;
+    }
+
+    NutritionPageLoaded.showNutritionPage(
```
```suggestion
    await NutritionPageLoaded.showNutritionPage(
```

monsieurtanuki reviewed (approved):
Thank you @Sudhanva-Nadiger, that's very good now!

monsieurtanuki committed:
Update packages/smooth_app/lib/pages/product/attribute_first_row_helper.dart

monsieurtanuki commented:
@g123k Please consider approving this excellent PR, as the merge button is blocked by your long-time-ago change request.

monsieurtanuki commented:
@g123k I didn't know I had the power to dismiss change requests. Given the context it looked like an appropriate action.
No hard feelings ;)

g123k commented:
> @g123k I didn't know I had the power to dismiss change requests. Given the context it looked like an appropriate action. No hard feelings ;)

No problem!

Sudhanva-Nadiger commented:
I havent added the Expand word to the translation ! there is one more pr continuation of this pr i am working on it ! i will add that in the next pr !
